src/org/opensolaris/opengrok/configuration/Group.java

```diff
@@ -0,0 +1,198 @@
+    }
+
+    public void addProject(Project p) {
+        this.projects.add(p);
+    }
+
+    public void addRepository(Project p) {
+        this.repositories.add(p);
+    }
+
+    public List<Project> getRepositories() {
+        return repositories;
+    }
+
+    /**
+     * @return List of all projects and repositories in this group or subgroups
+     */
+    public List<Project> getAllProjects() {
+        List<Project> p = new ArrayList<Project>(projects);
+        p.addAll(repositories);
+        for (Group g : getGroups()) {
+            p.addAll(g.getAllProjects());
+        }
+        sort(p, Project.COMPARATOR);
+        return p;
+    }
+
+    public void setGroups(List<Group> groups) {
+        this.groups = groups;
+    }
+
+    public void setProjects(List<Project> projects) {
+        this.projects = projects;
+    }
+
+    public void setRepositories(List<Project> repositories) {
+        this.repositories = repositories;
+    }
+
+    public List<Group> getGroups() {
+        return groups;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public String getPattern() {
+        return pattern;
+    }
+
+    public void setPattern(String pattern) {
+        this.pattern = pattern;
+    }
+
+    public int getFlag() {
+        return flag;
+    }
+
+    public void setFlag(int flag) {
+        this.flag = flag;
+    }
+
+    /**
+     * Finds recursively a group where project's description matches the group's
+     * pattern.
+     *
+     * First it tries subgroups and after that it tries itself.
+     *
+     * @param p project
+     * @return Group where the project belongs to
+     */
+    public Group match(Project p) {
+        Group subgroup = this;
+        for (Group g : this.groups) {
+            if ((subgroup = g.match(p)) != null) {
+                return subgroup;
+            }
+        }
+        if (Pattern.matches("(" + this.pattern + ")", p.getDescription())) {
+            return this;
+        }
+        return null;
+    }
+
+    /**
+     * Performs a deep copy of subgroups but leaves projects and repositories
+     * empty
+     *
+     * @return Group clone
+     */
+    public Group copy() {
+        Group g = new Group();
+        g.name = this.name;
+        g.pattern = this.pattern;
+        g.flag = this.flag;
+        for (Group x : groups) {
+            g.groups.add(x.copy());
+        }
+        return g;
+    }
+
+    /**
+     * Recursively checks if group is valid in the means of unique subgroup
+     * names along all group's children's siblings.
+     *
+     * Same names do not violate this condition if they are in different
+     * branches. Same names do not violate this condition if they are in the
+     * same branch but they are not siblings
+     *
+     * @param g Group
+     * @return true if it satisfies conditions above
+     */
+    public static boolean isValid(Group g) {
```

Comment:
I don't get what do we try to solve with this :)
But it also seems too complicated for me and that doesn't feel OK